src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,210 @@
+                            }
+                        }
+                        catch (IOException exception) when (retryCount < 3)
+                        {
+                            Console.WriteLine($"Exception caught: {exception.Message}, on retry count {retryCount}, Retrying\n");
+                            retryCount++;
+                        }
+                    }               
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount,
+                    CancellationToken = cancellationToken,
+                });
+
+            batchItemsBlock.LinkTo(fetchStream, new DataflowLinkOptions() { PropagateCompletion = true });
+            fetchStream.LinkTo(downloadBlock, new DataflowLinkOptions() { PropagateCompletion = true });
+
+            await batchItemsBlock.SendAllAndCompleteAsync(fileItems, downloadBlock, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<Stream> DownloadFileFromContainerAsync(Tuple<long, string> containerIdAndRoot, Guid scopeIdentifier, FileContainerHttpClient containerClient, FileContainerItem item, string targetPath, CancellationToken cancellationToken)
+        {
+            Stream responseStream = await AsyncHttpRetryHelper.InvokeAsync(
+                async () =>
+                {
+                    Stream internalResponseStream = await containerClient.DownloadFileAsync(containerIdAndRoot.Item1, item.Path, cancellationToken, scopeIdentifier).ConfigureAwait(false);
+                    return internalResponseStream;
+                },
+                maxRetries: 5,
+                cancellationToken: cancellationToken,
+                tracer: this.tracer,
+                continueOnCapturedContext: false
+                );
+
+            return responseStream;
+        }
+
+        private string ResolveTargetPath(string rootPath, FileContainerItem item, string artifactName)
+        {
+            var index = artifactName.Length;
+            var itemPathWithoutDirectoryPrefix = (index != -1 && index < item.Path.Length) ? item.Path.Substring(index + 1) : string.Empty;
+            var dirName = Path.GetDirectoryName(item.Path);
+            var absolutePath = Path.Combine(rootPath, itemPathWithoutDirectoryPrefix);
+            return absolutePath;
+        }
+
+        private List<FileContainerItem> GetFilteredItems(List<FileContainerItem> items, IEnumerable<Func<string, bool>> minimatchFuncs)
+        {
+            List<FileContainerItem> filteredItems = new List<FileContainerItem>();
+            foreach(FileContainerItem item in items)
+            {
+                int index = item.Path.IndexOf('/');
+                // trim the leading slash from the item paths
+                if (minimatchFuncs.Any(match => match(item.Path.Substring(index+1))))
```

Comment:
you can use string.Trim('/')

Reply:
Well, Trim is not exactly doing what is required, I think the comment is misleading, let me remove that.